Add logical types. Minor improvements.

siilike opened the pull request:
What's new:
* logical types (see below);
* immutable Buffers from pointers;
* minor improvements to the Collections class.

From the documentation:

> Types 0xee-0xef are used for tagging of values to implement logical
> types.
> 
> For example, if type 0x1c did not exist, the database driver could
> serialize a timestamp object (Date in JavaScript, Instant in Java, etc)
> into a Unix timestamp, a 64-bit integer. Assuming the lack of schema,
> upon deserialization it would not be possible to tell an integer from
> a timestamp and deserialize the value accordingly.
> 
> Type tagging resolves this by attaching an integer tag to values that
> can then be read when deserializing the value, e.g. that tag=1 is a
> timestamp and the relevant timestamp class should be used.
> 
> The tag values are specified separately and applications can also
> specify their own to have the database driver deserialize their specific
> data types into the appropriate classes (including models).
> 
> Essentially this is object-relational mapping for parts of documents.
> 
> The format of the type is:
> 
>     0xee
>     TAG in 1 byte
>     sub VPack value
> 
> or
> 
>     0xef
>     TAG in 8 bytes
>     sub VPack value

Tagging is implemented transparently -- operations on a Slice are unaffected and performed on the value of the tagged type, not on the type itself.

Looking forward to having this merged quickly.

graetzer commented:
can you give us a rational for implementing this into velocypack ? I feel this is a sbustantial change and we need to understand this better

graetzer commented:
seems to me like the application should know the schema ahead of time, no ?

graetzer commented on include/velocypack/Buffer.h:
```diff
@@ -49,6 +49,8 @@ class Buffer {
     initWithNone();
   }
 
+  Buffer(char *data, uint64_t len) noexcept : _buffer((unsigned char*) data), _capacity(len), _size(len), immutable(true) {}
```
why do you need an immutable Builder ? Kind of seems like a contradiction in itself

siilike commented:
The use case for this is applications where schema is not explicitly defined or is defined only partially, so the database driver will need additional information to deserialize the values properly.

The timestamp issue given above is a classical example. It can be solved either by having a dedicated type (like VelocyPack has) or by having a tag that signals that the integer that has been stored is a timestamp. The latter way also allows defining application-specific types and types that would not be very useful at a format level (like an age or a height type).

This is highly useful in simpler applications where one can get away with not explicitly specifying the schema and different abstractions, basically the MongoDB use case. But it assumes that what goes into the database comes out of there with the exact same classes, otherwise it will get very clumsy.

Regarding immutable buffers, I checked and turns out I don't have a use case for them any more, not sure what exactly they were for. I will update the pull request accordingly.

siilike commented:
Any updates on this?

jsteemann commented:
@siilike : sorry, all blame is on me, because I forgot to have a look at this PR.
I will get back within the next 24 hours.

siilike commented:
Great, there are some changes still missing to make it work with the latest changes in master.

siilike commented:
Now it is all updated. The only downside of the current implementation is that Slice supports transparency for only one level of tags, otherwise every access to _start would require iteration.

jsteemann commented on include/velocypack/Builder.h:
```diff
@@ -546,22 +613,26 @@ class Builder {
   template <typename T>
-  uint8_t* addInternal(char const* attrName, std::size_t attrLength, T const& sub) {
+  uint8_t* addInternal(char const* attrName, std::size_t attrLength, uint64_t tag, T const& sub) {
+    if(attrName == nullptr) {
```
This check adds a branch to a hot function, so I would like to avoid it.
Is the PR going to add calls to `addInternal` with an `attrName` argument which is a nullptr? In this case, these calls should be fixed instead.

siilike replied:
No.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
+  int64_t getTagId() const {
+    if(isTagged()) {
+      if(*_start == 0xee) {
+        return readIntegerFixed<ValueLength, 1>(_start + 1);
```
```suggestion
        return readIntegerFixed<uint64_t, 1>(_start + 1);
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
-  // pointer to the head byte
-  constexpr uint8_t const* start() const noexcept { return _start; }
-  
+  int64_t getTagId() const {
```
```suggestion
  uint64_t getTagId() const {
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
+      if(*_start == 0xee) {
+        return readIntegerFixed<ValueLength, 1>(_start + 1);
+      } else if(*_start == 0xef) {
+        return readIntegerFixed<ValueLength, 8>(_start + 1);
```
```suggestion
        return readIntegerFixed<uint64_t, 8>(_start + 1);
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
+  }
+
+  // pointer to the head byte, excluding possible tags
+  constexpr uint8_t const* start() const noexcept {
```
The `noexcept` is now misleading, as `tagOffset()` is potentially throwing.
I dislike removing the `noexcept` specifier from `start()`, as it is a very central and hot function, which is used in so many places in the library itself and in code that uses the library.

Additionally, this modification changes the behavior of `start()` so that it returns the VPack data without the tag. That means that embedders that use `start()` will lose all the tag information unless all embedder code is changed.
I think it should be the other way round: the behavior of `start()` should not be changed by this PR, instead there should be an additional (new) method that returns the VPack data without the tag. That will allow all embedder code to work the same way as before, as do a decoding of tags in the application, or, if not desired, call the new method to get the raw VPack data without the data.
But in any case, the behavior of `start()` should not be changed for downwards-compatibility reasons.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
-  
-  uint8_t const* begin() noexcept { return _start; }
+  // value of the head byte (excluding tags)
+  constexpr inline uint8_t head() const noexcept { return *start(); }
```
The changes to `start()` will lead to this method not being `noexcept` anymore.
IMHO this should be reverted to just return `*_start` again, as this will ensure the old behavior that embedders may rely on.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
+  constexpr inline uint8_t rawHead() const noexcept { return *rawStart(); }
 
-  uint8_t const* begin() const noexcept { return _start; }
+  uint8_t const* begin() noexcept { return start(); }
```
The changes to `start()` will lead to this method not being `noexcept` anymore.
IMHO this should be reverted to just return `_start` again, as this will ensure the old behavior that embedders may rely on.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
+  uint8_t const* begin() noexcept { return start(); }
 
-  uint8_t const* end() { return _start + byteSize(); }
+  constexpr uint8_t const* begin() const noexcept { return start(); }
```
Why have two different `begin()` methods now? I think there should only be one, and it should be const.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
+  constexpr uint8_t const* begin() const noexcept { return start(); }
 
-  uint8_t const* end() const { return _start + byteSize(); }
+  uint8_t const* end() { return start() + byteSize(); }
```
IMHO this should be reverted to just return `_start + byteSize()` again, as this will ensure the old behavior that embedders may rely on.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
+
+  // pointer to the head byte, including possible tags
+  constexpr uint8_t const* rawStart() const noexcept {
+    return _start;
```
what about adding changing this to `_start + getTagOffset()` instead, so `rawStart()` can be used to get the VPack value without tag information? This also means the `noexcept` specifier needs to be removed here.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
+  }
+
+  constexpr uint8_t tagOffset() const {
+    return tagOffset(rawStart());
```
```suggestion
    return tagOffset(_start);
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
   template <typename T>
   T const* startAs() const {
-    return reinterpret_cast<T const*>(_start);
+    return reinterpret_cast<T const*>(start());
```
IMHO this should be reverted to just return (the converted pointer to) `_start` again, as this will ensure the old behavior that embedders may rely on.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,38 +117,70 @@ class Slice {
+  uint8_t const* end() { return start() + byteSize(); }
 
-  // get the type for the slice
+  uint8_t const* end() const { return start() + byteSize(); }
```
I don't think there is a need for a second `end()` implementation. There should only be one, and it should be const.

siilike replied:
Yes, that's probably a merging issue.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -181,9 +219,14 @@ class Slice {
   }
 
-  // check if slice is of the specified type
+  // check if slice is of the specified type (excluding tags)
```
```suggestion
  // check if slice is of the specified type (including tags)
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -181,9 +219,14 @@ class Slice {
+  // check if slice is of the specified type (excluding tags)
   constexpr inline bool isType(ValueType t) const noexcept {
-    return SliceStaticData::TypeMap[*_start] == t;
+    return SliceStaticData::TypeMap[*start()] == t;
```
IMHO this should be reverted to just return the type based on `_start` again, as this will ensure the old behavior that embedders may rely on.
```suggestion
    return SliceStaticData::TypeMap[_start] == t;
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -181,9 +219,14 @@ class Slice {
+    return SliceStaticData::TypeMap[*start()] == t;
+  }
+
+  // check if slice is of the specified type (including tags)
```
```suggestion
  // check if slice is of the specified type (excluding tags)
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -249,6 +292,9 @@ class Slice {
   constexpr bool isCustom() const noexcept { return isType(ValueType::Custom); }
 
+  // check if slice is a Tagged type
+  constexpr bool isTagged() const noexcept { return isRawType(ValueType::Tagged); }
```
```suggestion
  constexpr bool isTagged() const noexcept { return isType(ValueType::Tagged); }
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -748,133 +794,40 @@ class Slice {
   }
 
-  // get the total byte size for the slice, including the head byte
+  // get the total byte size for the slice, including the head byte, excluding tags
```
```suggestion
  // get the total byte size for the slice, including the head byte, including tags
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -748,133 +794,40 @@ class Slice {
-  // get the total byte size for the slice, including the head byte
+  // get the total byte size for the slice, including the head byte, excluding tags
   ValueLength byteSize() const {
-    auto const h = head();
```
IMHO this change should be reverted, because it changes the behavior for all embedders and will cause subtly out-of-bounds errors for applications that want to use tagged values.
Of course a case for `Tagged` needs to be added to the old implementation to accommodate for tagged values.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -748,133 +794,40 @@ class Slice {
+  }
 
-    throw Exception(Exception::InternalError, "Invalid type for byteSize()");
+  // get the total byte size for the slice, including the head byte, including tags
```
```suggestion
  // get the total byte size for the slice, including the head byte, excluding tags
```

jsteemann commented on include/velocypack/velocypack-common.h:
```diff
@@ -94,8 +94,10 @@
 #ifdef VELOCYPACK_XXHASH
 // forward for XXH64 function declared elsewhere
 extern "C" unsigned long long XXH64(void const*, std::size_t, unsigned long long);
+extern "C" unsigned int XXH32(const void* input, size_t len, unsigned int seed);
```
```suggestion
extern "C" unsigned int XXH32(void const* input, std::size_t len, unsigned int seed);
```

jsteemann commented on src/Collection.cpp:
```diff
@@ -137,6 +139,28 @@ bool Collection::contains(Slice const& slice, Slice const& other) {
 }
 
+bool Collection::containsObject(Slice const& slice, Slice const& other) {
+  ObjectIterator it(other);
```
```suggestion
  ObjectIterator it(other, true);
```

jsteemann commented on src/Collection.cpp:
```diff
@@ -137,6 +139,28 @@ bool Collection::contains(Slice const& slice, Slice const& other) {
   return false;
 }
 
+bool Collection::containsObject(Slice const& slice, Slice const& other) {
```
should this be renamed to `containsSubobject`?

jsteemann commented on include/velocypack/Collection.h:
```diff
@@ -80,6 +82,8 @@ class Collection {
 
   static bool contains(Slice const& slice, Slice const& other);
 
+  static bool containsObject(Slice const& slice, Slice const& other);
```
should this be renamed to `containsSubobject`?

siilike replied:
I checked and I actually don't have any use for this method, so I removed it for now.

I was also thinking that it is rather confusing to have array and object methods in the same class as they are not compatible, e.g. it is not possible to filter() or concat() an object, so it would make sense to either add support for objects to all methods or create an entire separate class for handling the same operations for objects.

jsteemann commented:
In general, I think the idea of tags makes sense for some applications.
However, this PR changes the behavior of several public methods in `velocypack::Slice`, so when existing applications start using tags, all existing calls in these applications to `Slice::start()`, `Slice::byteSize()` etc. will need to be changed.
Instead, the public methods in `Slice` should remain downwards-compatible. Otherwise, only the VPack data without tags will make it into client applications, and client applications have no chance of using the tags for anything.
Additionally, I now a few embedders that use the combination of `start()` and `byteSize()` and feed that into memcpy. If the behavior of `start()` and `byteSize()` however changes, then all tags will instantly be lost unless all embedder code is changed.

jsteemann commented:
And finally, we **must** add tests for tagged values and for the new method in `Collections` before we can this PR.

jsteemann commented:
if it helps, I am also happy to split up this multi-purpose PR into multiple small single-purpose PRs. That way we can already merge the following bits quickly, without having to wait for the bigger tags change:
- the `Collection`-related additions (plus tests)
- the 32bit hash function (plus tests)
- CMakeLists.txt changes
WDYT?

siilike commented:
Thank you for reviewing.

The idea of making the tags transparent by default was that it makes working with the values a lot simpler. Otherwise there would need to be a method that returns the value if tagged and otherwise the object itself which would clutter the code and be an easy source of errors if the value initially isn't tagged, but becomes at some point.

But the important part is that is* and get* methods apply to the value of the tagged value, not the tagged value itself, so start(), end() and others can be reverted to apply to the whole value. This way we'll have enough compatibility with existing solutions and convenient access to values.

siilike commented:
I removed the exception from Slice.tagOffset, because the only time when there can be an error is when a new tagged type is added to SliceStaticData.TypeMap and that method is not updated. This removes all the headache of noexcept.

I now updated the transparency logic as follows:
  - methods that deal with the pointer and length keep the former behaviour and have a relevant value* method: start, startAs, head, begin, end, byteSize;
  - in methods that deal with the value tags are transparent and some have a relevant raw* method: type, typeName, is*, get*, copy*, has*, etc.

If that is fine I'll also update Slice.cpp and any references elsewhere.

siilike commented:
ping

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -526,7 +582,7 @@ class Slice {
   // an External, recursive version
   Slice resolveExternals() const {
-    uint8_t const* current = _start;
+    uint8_t const* current = valueStart();
```
I think this should still operate on `start()` and not on `valueStart()`.
If a client application needs to peek into the value behind the tag, I think it should explicitly resolve the tag.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -621,45 +677,45 @@ class Slice {
       throw Exception(Exception::InvalidValueType, "Expecting type UTCDate");
     }
-    uint64_t v = readIntegerFixed<uint64_t, sizeof(uint64_t)>(_start + 1);
+    uint64_t v = readIntegerFixed<uint64_t, sizeof(uint64_t)>(valueStart() + 1);
```
I think this should still operate on `start()` and not on `valueStart()`.
If a client application needs to peek into the value behind the tag, I think it should explicitly resolve the tag.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -621,45 +677,45 @@ class Slice {
       // short UTF-8 String
       length = h - 0x40;
-      return reinterpret_cast<char const*>(_start + 1);
+      return reinterpret_cast<char const*>(valueStart() + 1);
```
I think this should still operate on `start()` and not on `valueStart()`.
If a client application needs to peek into the value behind the tag, I think it should explicitly resolve the tag.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -621,45 +677,45 @@ class Slice {
   // return the value for a String object
   char const* getString(ValueLength& length) const {
-    uint8_t const h = head();
+    uint8_t const h = valueHead();
```
I think this should still operate on `head()` and not on `valueHead()`.
If a client application needs to peek into the value behind the tag, I think it should explicitly resolve the tag.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -621,45 +677,45 @@ class Slice {
     if (h == 0xbf) {
       // long UTF-8 String
-      length = readIntegerFixed<ValueLength, 8>(_start + 1);
+      length = readIntegerFixed<ValueLength, 8>(valueStart() + 1);
```
I think this should still operate on `start()` and not on `valueStart()`.
If a client application needs to peek into the value behind the tag, I think it should explicitly resolve the tag.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -621,45 +677,45 @@ class Slice {
+      length = readIntegerFixed<ValueLength, 8>(valueStart() + 1);
       checkOverflow(length);
-      return reinterpret_cast<char const*>(_start + 1 + 8);
+      return reinterpret_cast<char const*>(valueStart() + 1 + 8);
```
I think this should still operate on `start()` and not on `valueStart()`.
If a client application needs to peek into the value behind the tag, I think it should explicitly resolve the tag.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -621,45 +677,45 @@ class Slice {
   
   char const* getStringUnchecked(ValueLength& length) const noexcept {
-    uint8_t const h = head();
+    uint8_t const h = valueHead();
```
I think this should still operate on `head()` and not on `valueHead()`.
If a client application needs to peek into the value behind the tag, I think it should explicitly resolve the tag.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -621,45 +677,45 @@ class Slice {
       // short UTF-8 String
       length = h - 0x40;
-      return reinterpret_cast<char const*>(_start + 1);
+      return reinterpret_cast<char const*>(valueStart() + 1);
```
I feel I am repeating the same sentence here over and over :smile:
The same argument applies for most of the following functions.

siilike commented:
Pushed the third take on this:
- Slice has no behavioural changes, only tags support added;
- created new ValueSlice class where tags are transparent;
- added support for multiple tags -- initially I skipped this as I did not want to introduce iteration everywhere, but it seems lousy not to implement it when technically it is possible to have multiple tags.
If this is fine I'll also update Builder.

siilike commented:
Ping. Can't write tests before we've agreed on Slice.

siilike commented:
Ping. I really don't want to fork this project and have yet another data format.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -58,6 +59,7 @@ class Slice {
   friend class ArrayIterator;
   friend class ObjectIterator;
 
+  // This variable should be accessed through the start() method
```
I don't think this comment is justified. The `start()` method simply returns `_start` as is, and `_start` is a private member. So I don't see a great need for the VPackSlice class to use the `start()` method to access `_start`.

siilike replied:
It should, because otherwise ValueSlice would not work. ValueSlice overrides start() to point to the actual value, so that all inherited methods will also use it. Otherwise I would have to duplicate everything that uses _start in ValueSlice which is unreasonable.

jsteemann replied:
Ok, I overlooked this. Makes sense. If you could add that to the comment it would help me (and potentially others) to understand the implications of using `_start` vs. `start()`.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -117,36 +122,108 @@ class Slice {
+             );
+  }
+
+  std::list<uint64_t> getTags() const {
```
I think an `std::vector<uint64_t>` would be more lightweight here.
`std::list` needs to allocate each member seperately, whereas a vector doesn't need allocations per element (effectively it will do a few allocations for the first few elements, but afterwards it will reallocate a lot less).
Apart from that, using vector here allows removing the `<list>` include from the header and reduces dependencies.

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -748,133 +835,41 @@ class Slice {
-    throw Exception(Exception::InternalError, "Invalid type for byteSize()");
+  // get the total byte size for the slice, including the head byte, excluding tags
+  ValueLength valueByteSize() const {
+    return byteSize(start());
```
`byteSize()` and `valueByteSize()` perform the same task, so one of them (`byteSize()`) would be enough.
What's more confusing is that there is now a `value()` function that skips tags, but `valueByteSize()` doesn't.
So I suggest we either remove `valueByteSize()` or adjust its behavior so that it skips tags too.

siilike replied:
That looks like a typo, it should be valueStart().

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -1100,9 +1116,135 @@ class Slice {
     return value;
   }
+
+  constexpr uint8_t tagOffset(uint8_t const* _start) const noexcept {
```
The function parameter names shouldn't start with an underscore, to avoid confusion with the instance variable of the same name.
```suggestion
  constexpr uint8_t tagOffset(uint8_t const* start) const noexcept {
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -1100,9 +1116,135 @@ class Slice {
   }
+
+  constexpr uint8_t tagOffset(uint8_t const* _start) const noexcept {
+    return SliceStaticData::TypeMap[*_start] == ValueType::Tagged ? (*_start == 0xee ? 2 : (*_start == 0xef ? 9 : /* error */ 0)) : 0;
```
```suggestion
    return SliceStaticData::TypeMap[*start] == ValueType::Tagged ? (*start == 0xee ? 2 : (*start == 0xef ? 9 : /* error */ 0)) : 0;
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -1100,9 +1116,135 @@ class Slice {
+    return SliceStaticData::TypeMap[*_start] == ValueType::Tagged ? (*_start == 0xee ? 2 : (*_start == 0xef ? 9 : /* error */ 0)) : 0;
+  }
+
+  uint8_t tagsOffset(uint8_t const* _start) const noexcept {
```
same here
```suggestion
  uint8_t tagsOffset(uint8_t const* start) const noexcept {
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -1100,9 +1116,135 @@ class Slice {
+  uint8_t tagsOffset(uint8_t const* _start) const noexcept {
+    uint8_t ret = 0;
+
+    while(SliceStaticData::TypeMap[*_start] == ValueType::Tagged) {
```
```suggestion
    while(SliceStaticData::TypeMap[*start] == ValueType::Tagged) {
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -1100,9 +1116,135 @@ class Slice {
+    uint8_t ret = 0;
+
+    while(SliceStaticData::TypeMap[*_start] == ValueType::Tagged) {
+      uint8_t offset = tagOffset(_start);
```
```suggestion
      uint8_t offset = tagOffset(start);
```

jsteemann commented on include/velocypack/Slice.h:
```diff
@@ -1100,9 +1116,135 @@ class Slice {
+    while(SliceStaticData::TypeMap[*_start] == ValueType::Tagged) {
+      uint8_t offset = tagOffset(_start);
+      ret += offset;
+      _start += offset;
```
```suggestion
      start += offset;
```

jsteemann commented on include/velocypack/ValueSlice.h:
```diff
@@ -0,0 +1,160 @@
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Max Neunhoeffer
+/// @author Jan Steemann
```
You can put in your name instead. It's your code. :smile:

siilike replied:
Yes, I didn't notice when copying the header.

jsteemann commented on include/velocypack/ValueSlice.h:
```diff
@@ -0,0 +1,160 @@
+#include <iosfwd>
+#include <algorithm>
+#include <functional>
+#include <type_traits>
```
I think most of these headers aren't needed by this file.

jsteemann commented on src/Dumper.cpp:
```diff
@@ -413,11 +441,39 @@ void Dumper::dumpValue(Slice const* slice, Slice const* base) {
+          _sink->push_back('N');
+          _sink->push_back('a');
+          _sink->push_back('N');
+          _sink->push_back('"');
```
Looks like `_sink->append("\"NaN\"", 5);` would have been easier.

jsteemann commented on src/Dumper.cpp:
```diff
@@ -413,11 +441,39 @@ void Dumper::dumpValue(Slice const* slice, Slice const* base) {
+          _sink->push_back('i');
+          _sink->push_back('t');
+          _sink->push_back('y');
+          _sink->push_back('"');
```
Can be turned into `_sink->append("Infinity\"", 9);`.

siilike commented on include/velocypack/Builder.h:
```diff
@@ -546,22 +647,29 @@ class Builder {
+  uint8_t* addInternal(char const* attrName, std::size_t attrLength, uint64_t tag, T const& sub) {
+    // When attrName is null add an array element instead -- simplifies code
+    // that builds both objects and arrays
+    if(attrName == nullptr) {
```
The use case I have for this is when same code can generate both arrays and objects. I have a function for type conversion that has over 20 calls to add(Slice), add(Value) and add(ValuePair) and there it is very convenient to just pass a null key to signal that an array element should be added instead. Otherwise I would have to duplicate those calls.

But if you find that it is not reasonable I can live with one ugly function.

jsteemann replied:
I would prefer that function only to be called with with valid `attrName`s.

siilike commented:
Great, thank you, will write tests now.

siilike commented:
Turns out this overriding idea does not work. The start() method would have to be virtual which would mean the removal of constexpr and sizeof(arangodb::velocypack::Slice) != sizeof(void*).

siilike commented:
OK, I got rid of ValueSlice for now as I can temporarily give up the convenience until I find the time to resolve this. So, please review the tests and merge if everything is fine.

OmarAyo commented:
Hi @siilike 

Many thanks for contributing. We are obliged by law to have a signed scanned copy of a Contributor License Agreement (CLA) to be able to accept pull requests to the official ArangoDB repository - 

We use an Apache 2 CLA for ArangoDB and its companion projects, which can be found here: https://www.arangodb.com/documents/cla.pdf
Just fill in the form, sign and send a scanned copy over to cla@arangodb.com

Thank you

OmarAyo commented:
CLA available

siilike commented:
Ping

siilike commented:
Still waiting.

siilike commented:
Yet another week has gone by.

jsteemann commented:
@siilike : Hi, sorry for the delay. I am currently busy with lots of other things as we are preparing a release. I will make time to look into this on either Wed or Thu this week. Sorry again!

siilike commented:
Great, thanks.

I made a separate pull request for the specification change, so we can continue with arangodb/java-velocypack#15 in the meantime.

jsteemann commented on src/Dumper.cpp:
```diff
@@ -444,9 +489,40 @@ void Dumper::dumpValue(Slice const* slice, Slice const* base) {
+
+    case ValueType::UTCDate: {
+      if(options->datesAsIntegers) {
+        appendUInt(slice->getUTCDate());
```
AFAIK `Slice::getUTCDate()` returns an `int64_t`, so this should be changed to
```suggestion
        appendInt(slice->getUTCDate());
```

jsteemann commented on include/velocypack/velocypack-memory.h:
```diff
@@ -0,0 +1,50 @@
+
+extern "C" {
+
+extern void* vmalloc(std::size_t size);
```
Using a very short name such `vmalloc` is a bit too risky for my gusto, because it may easily be defined by some other component (e.g. another malloc implementation or a system header).

Can these functions be renamed to `velocypack_malloc`, `velocypack_realloc` and `velocypack_free`?
This will make it unambiguous and not clash with anything else.

siilike replied:
Sure, will do.

jsteemann commented on src/Builder.cpp:
```diff
@@ -52,12 +52,18 @@ struct SortEntry {
 constexpr size_t minSortEntriesAllocation = 32;
 
+
+#ifndef NO_THREADLOCALS
```
I think it is better to use macro names prefixed with the library name here, again to make them unambiguous.
```suggestion
#ifndef VELOCYPACK_NO_THREADLOCALS
```
Additionally just wondering what the problem is with requiring thread-locals. Is there any issue with them?

siilike replied:
The issue is that they do not play well with memory contexts, they will point to freed memory once the memory context gets destroyed.

jsteemann commented:
@siilike : thanks for your patience! I left a few comments, but apart from these small issues I think it is good to go now.

siilike commented:
Great, thank you. I made the fixes.

jsteemann commented:
Thanks a lot for the patch!
